Initial buzzfeed-sso chart - Oauth over your Kubernetes services

darioblanco opened the pull request:
Signed-off-by: Dario Blanco <dblanco@minddoc.de>

#### What this PR does / why we need it:

Having a cluster-global oauth service that protects other services even from different namespaces is becoming a common use case. This chart offers a solution for easily securing services (like monitoring dashboards) using SSO and a central place where to put your credentials.

BuzzFeed has open sourced a very interesting SSO solution that seems quite promising, as the current oauth2_proxy is not maintained anymore. However, they have no helm chart and could be a bit challenging to set it up on your own. This branch hopes to abstract all those days of work into a very easy setup.

#### Special notes for your reviewer:

For now I give support to Google SSO, as this is our production use case, and will continue to do so. However, I think I have set a basic abstraction so other providers would be easily added, if somebody else wants to contribute.

#### Checklist
- [x] [DCO](https://www.helm.sh/blog/helm-dco/index.html) signed
- [x] Variables are documented in the README.md

Commit:
Initial sso chart

Signed-off-by: Dario Blanco <dblanco@minddoc.de>

Commit:
Rename to buzzfeed-sso

Signed-off-by: Dario Blanco <dblanco@minddoc.de>

Commit:
Remove auth code secret from proxy

Signed-off-by: Dario Blanco <dblanco@minddoc.de>

Commit:
Make tls templating more flexible

Signed-off-by: Dario Blanco <dblanco@minddoc.de>

Commit:
Remove proxy code secret from the configuration

Signed-off-by: Dario Blanco <dblanco@minddoc.de>

naseemkullah commented on stable/buzzfeed-sso/README.md:
````diff
@@ -0,0 +1,149 @@
+To install the chart with the release name `my-release`:
+
+```bash
+$ helm install --name my-release stable/sso
````
Please adjust chart name here and anywhere else in the readme since the settled-on chart name is `buzzfeed-sso` and not simply `sso`

weeco commented on stable/buzzfeed-sso/templates/proxy-deployment.yaml:
```diff
@@ -0,0 +1,100 @@
+            - name: CLUSTER
+              value: dev
+            - name: VIRTUAL_HOST
+              value: "{{ .Values.proxy.virtualHost }}"
```
```suggestion
              value: {{ .Values.proxy.virtualHost | quote }}
```

Commit:
Use buzzfeed-sso in README

Signed-off-by: Dario Blanco <dblanco@minddoc.de>

Commit:
Use quote filter in proxy virtualHost variable

Signed-off-by: Dario Blanco <dblanco@minddoc.de>

Commit:
Set CLUSTER in proxy as configurable and dev by default

Signed-off-by: Dario Blanco <dblanco@minddoc.de>

Commit:
Quote all templated environment variables

Signed-off-by: Dario Blanco <dblanco@minddoc.de>

Commit:
Set cookie secure to true by default

Signed-off-by: Dario Blanco <dblanco@minddoc.de>

Commit:
Allow definition of a custom secret key
This will skip secret creation via helm

Signed-off-by: Dario Blanco <dblanco@minddoc.de>

asmirnou commented on stable/buzzfeed-sso/templates/ingress.yaml:
```diff
@@ -0,0 +1,46 @@
+  tls:
+    - hosts:
+        - {{ $authDomain }}
+      secretName: sso-auth-tls-secret
```
This secret has to be set from `values`, hasn't it?

darioblanco replied:
I guess we can make it dynamic, though an auth domain is always required. Do you think is essential for this PR or something that can be implemented later?

asmirnou replied:
In my case TLS secret is same for auth and proxy hosts and it already exists, so when working with predefined secrets, this line should be dynamic.
In my copy I added `tls.secretName` for auth to the list of hosts in ingress, but excluded auth domain from the loop in configmap template in order to not have it in `upstream_configs.yml`. This eliminates the necessity of another value for just auth secret and make values file more consistent.

darioblanco replied:
I have added a `tls.secretName` configuration option, so the secret name can be dynamically changed

ptrus commented on stable/buzzfeed-sso/templates/configmap.yaml:
```diff
@@ -0,0 +1,17 @@
+data:
+  upstream_configs.yml: |-
+  {{- range .Values.ingress.hosts }}
+    - service: {{ .serviceName }}
```
I think it would make sense to add support for configuring other 'upstream_config' parameters as well, see https://github.com/buzzfeed/sso/blob/f437f237ac977201f15868601c9bc0e9dff11f40/docs/sso_config.md#proxy-config

darioblanco replied:
Good point. Even if we will need to repeat the domain again, I have separated the `ingress.hosts` configuration from the `upstreams`. Now we have an `upstreams` list of configurations that exactly fits with the url you just posted.

This should give complete flexibility about how the upstreams are configured

Commit:
Make auth ingress tls secret name configurable

Signed-off-by: Dario Blanco <dblanco@minddoc.de>

Commit:
Separate upstream configuration from ingress

Signed-off-by: Dario Blanco <dblanco@minddoc.de>

Commit:
Update image to v1.1.0
Adds a new optional environment variable for the proxy that allows split dns deployments

Signed-off-by: Dario Blanco <dblanco@minddoc.de>

Commit:
Move chart to incubator

Signed-off-by: Dario Blanco <dblanco@minddoc.de>

bskim45 commented on incubator/buzzfeed-sso/templates/auth-deployment.yaml:
```diff
@@ -0,0 +1,141 @@
+    release: {{ .Release.Name }}
+    heritage: {{ .Release.Service }}
+spec:
+  replicas: {{ .Values.proxy.replicaCount }}
```
.Values.auth.replicaCount ? 😅

darioblanco replied:
Oops, good catch

bskim45 commented on incubator/buzzfeed-sso/templates/auth-deployment.yaml:
```diff
@@ -0,0 +1,141 @@
+  name: {{ $fullName }}-auth
+  labels:
+    app: {{ $name }}-auth
+    chart: {{ template "buzzfeed-sso.chart" . }}-auth
```
The `app`, `chart` label should be the same across all chart components. How about using `component` label to distinguish them?

darioblanco replied:
Good point 👍 . I am using the `container` label now to distinguish them.

Commit:
Fix auth configuration variables

Signed-off-by: Dario Blanco <dblanco@minddoc.de>

Commit:
Use component label to distinguish between auth and proxy

Signed-off-by: Dario Blanco <dblanco@minddoc.de>

victornoel commented on incubator/buzzfeed-sso/values.yaml:
```diff
@@ -0,0 +1,98 @@
+    # ingress.kubernetes.io/ssl-redirect: "true"
+  hosts: []
+  #  - domain: mybackend.mydomain.foo
+  #  - path: /
```
I think the `-` should be removed here, no?

darioblanco replied:
Why? It can still be used in the ingress.

victornoel replied:
@darioblanco  If I'm not mistaken, the `hosts` field takes an array of objects that each contains `domain` and `path`, so it should look like this:
```
hosts:
  - domain: xxx
    path: xxxx
```

In the code above, `hosts` is an array of two elements, one containing `domain`, the other `path`, was that really on purpose?

darioblanco replied:
Ohhh now I get it. Even if I looked at it more than once I did not realize the mistake. Will fix it right away in the comments so is not misleading

victornoel commented on incubator/buzzfeed-sso/values.yaml:
```diff
@@ -0,0 +1,98 @@
+    type: ClusterIP
+    port: 80
+  # Generate these secrets with the command:
+  # 'openssl rand -base64 32 | head -c 32 | base64'
```
wouldn't it better to do the `b64enc` in the helm template?

darioblanco replied:
Those will be encoded in `secret.yaml`. How the secrets are managed outside the chart, in my opinion, should be controlled by the user.

victornoel replied:
@darioblanco my mistake, actually I didn't realise that SSO was requiring the secret to be encoded in base64 and then that it was encoded again in the secret resource in the template, you can ignore this comment :)

victornoel commented on incubator/buzzfeed-sso/README.md:
```diff
@@ -0,0 +1,183 @@
+`auth.tls` | tls configuration for central sso auth ingress. | `{ secretName: "sso-auth-tls-secret" }`
+`auth.customSecret` | the secret key to reuse (avoids secret creation via helm) | REQUIRED if `auth.secret` is not set
+`proxy.virtualHost` | wildcard domain for redirecting SSO to the backends | REQUIRED
+`proxy.providerUrlInternal` | url for split dns deployments |
```
@darioblanco I believe this should be better documented because I just can't understand how to use this :)

victornoel replied:
from what I gathered, it needs to be set to the internal url of the auth service (so it depends on the release name), if not, things won't work.
I wonder if it wouldn't be better to simply set it to the correct value in the template if `proxy.providerUrlInternal` is not provided by the user

darioblanco replied:
This is a new feature from the latest sso update. TBH I haven't tried it and the setting is there, plus there wasn't much documentation about it in the buzzfeed sso project.

Do you mean that the chart does not work? If you don't provide the `proxy.providerUrlInternal` it is just not set in the environment variables. Can you please explain a bit what is the problem here?

victornoel replied:
@darioblanco from my tests, without this setting I couldn't make the chart work: the proxy would refuse my requests even though I had been authenticated first by the auth service.
By setting `proxy.providerUrlInternal` to the internal name in the cluster of the auth service, it started to work, but maybe I made a mistake elsewhere also and thought that was the problem?

This is what I put in the template:
```
            - name: PROVIDER_URL_INTERNAL
              {{- if .Values.proxy.providerUrlInternal }}
              value: {{ .Values.proxy.providerUrlInternal | quote }}
              {{- else }}
              value: "http://{{ template "buzzfeed-sso.fullname" . }}-auth:80"
              {{- end }}
```

darioblanco replied:
Hmm, I can verify that it works without this setting perfectly. Even I applied the latest changes to the production cluster.
However, I have never tried to set the `PROVIDER_URL_INTERNAL`. It was added there as it was a new environment variable from the new buzzfeed version. As far as I know it works without it. And with it, it should work as well, as it is not doing anything different than what it is stated in buzzfeed's documentation.

Therefore I don't think it is necessary to put the else there.

victornoel commented on incubator/buzzfeed-sso/README.md:
```diff
@@ -0,0 +1,183 @@
+`proxy.secret` | secrets to be generated randomly with `openssl rand -base64 32 | head -c 32 | base64`. | REQUIRED if `proxy.customSecret` is not set
+`proxy.customSecret` | the secret key to reuse (avoids secret creation via helm) | REQUIRED if `proxy.secret` is not set
+`provider.google` | the Oauth provider to use (only Google support for now) | REQUIRED
+`provider.google.adminEmail` | the Google admin email | REQUIRED
```
@darioblanco it can be useful to have this value being optional. If not specified, we would skip everything related to the service account in the auth deployment for people who don't want to setup group access control

darioblanco replied:
Now the environment variable for the google admin email will be skipped if not specified. I did not know that it wasn't required, thanks for the feedback!

victornoel commented on incubator/buzzfeed-sso/templates/auth-deployment.yaml:
```diff
@@ -0,0 +1,144 @@
+  template:
+    metadata:
+      annotations:
+        checksum/config: {{ include (print $.Template.BasePath "/configmap.yaml") . | sha256sum }}
```
@darioblanco the checksum should also contain something about the secrets so that when they change, the pods are restarted.

darioblanco replied:
I am thinking in reworking the checksum behavior and it would be useful to allow the user to specify annotations (for instance, I use fluentbit and would like to skip some logs).

Will work on it on this branch

darioblanco replied:
Fixed. In addition, now it is possible to configure annotations. This is quite handy for me, as I have already pushed my `fluentbit.io/exclude: "true"` annotation and I have checked that it works in my production cluster

victornoel commented on incubator/buzzfeed-sso/templates/proxy-deployment.yaml:
```diff
@@ -0,0 +1,108 @@
+  template:
+    metadata:
+      annotations:
+        checksum/config: {{ include (print $.Template.BasePath "/configmap.yaml") . | sha256sum }}
```
@darioblanco the checksum should also contain something about the secrets so that when they change, the pods are restarted.

darioblanco replied:
I have added a `checksum/secret` annotation to the proxy and the auth pods

victornoel commented on incubator/buzzfeed-sso/templates/proxy-deployment.yaml:
```diff
@@ -0,0 +1,108 @@
+            - name: VIRTUAL_HOST
+              value: {{ .Values.proxy.virtualHost | quote }}
+          {{- if .Values.proxy.providerUrlInternal }}
+            - name: PROXY_PROVIDER_URL
```
@darioblanco this has been renamed PROVIDER_URL_INTERNAL apparently (see https://github.com/buzzfeed/sso/releases/tag/v1.1.0)

darioblanco replied:
Ouch, good catch

victornoel commented on incubator/buzzfeed-sso/templates/auth-deployment.yaml:
```diff
@@ -0,0 +1,144 @@
+              value: "true"
+            - name: CLUSTER
+              value: dev
+            - name: VIRTUAL_HOST
```
@darioblanco I believe this environment variable is only useful when using [nginx-proxy](https://github.com/jwilder/nginx-proxy) as in the docker-compose example https://github.com/buzzfeed/sso/blob/master/quickstart/docker-compose.yml

darioblanco replied:
In any case, I would keep it as it is. We set the `VIRTUAL_HOST` to `auth.domain`, which is also used for the ingress configuration and the `PROVIDER_URL` in the proxy.
Unless we know 100% it is useless for Kubernetes deployments (for instance, I use the `nginx-ingress` and I haven't tested if it works without it).

darioblanco replied:
It works, I am removing this as requested, so we can reduce complexity

victornoel commented on incubator/buzzfeed-sso/README.md:
```diff
@@ -0,0 +1,183 @@
+`auth.secret` | secrets to be generated randomly with `openssl rand -base64 32 | head -c 32`. | REQUIRED if `auth.customSecret` is not set
+`auth.tls` | tls configuration for central sso auth ingress. | `{ secretName: "sso-auth-tls-secret" }`
+`auth.customSecret` | the secret key to reuse (avoids secret creation via helm) | REQUIRED if `auth.secret` is not set
+`proxy.virtualHost` | wildcard domain for redirecting SSO to the backends | REQUIRED
```
@darioblanco I believe this is not needed (see https://github.com/helm/charts/pull/8157/files#r248581150)

darioblanco replied:
Indeed, it is not needed

Commit:
Make google admin email optional

Signed-off-by: Dario Blanco <dblanco@minddoc.de>

Commit:
Fix proxy url internal environment variable

Signed-off-by: Dario Blanco <dblanco@minddoc.de>

victornoel commented on incubator/buzzfeed-sso/templates/auth-deployment.yaml:
```diff
@@ -95,8 +95,10 @@ spec:
+          {{- if .adminEmail }}
             - name: GOOGLE_ADMIN_EMAIL
               value: {{ .adminEmail | quote }}
+          {{- end }}
```
@darioblanco I would also take inside the if the `GOOGLE_SERVICE_ACCOUNT_JSON` since it is useful only if the `adminEmail` is present. Same with the `google-service-account` volumeMounts.

victornoel replied:
(the whole point of making `adminEmail` optional is to avoid having to generate a service account when one does not want to use group based access control)

darioblanco replied:
Updated. If no admin email is given, the google provider will only work with the client id and client secret stored in `$googleSecret`

Commit:
Remove unnecessary virtual host

Signed-off-by: Dario Blanco <dblanco@minddoc.de>

Commit:
Add only google service account credentials when admin email is given

Signed-off-by: Dario Blanco <dblanco@minddoc.de>

Commit:
Make annotations configurable
It also adds the secret file to the checksum

Signed-off-by: Dario Blanco <dblanco@minddoc.de>

Commit:
Fix ingress hosts example

Signed-off-by: Dario Blanco <dblanco@minddoc.de>

Commit:
Allow installation with default values showing errors

Signed-off-by: Dario Blanco <dblanco@minddoc.de>

Commit:
Remove unusued virtualHost parameter

Signed-off-by: Dario Blanco <dblanco@minddoc.de>

Commit:
Add OWNERS file

Signed-off-by: Dario Blanco <dblanco@minddoc.de>
